lib/wallet/src/Cardano/Wallet/Read/Tx/CBOR.hs

```diff
@@ -1,100 +1,123 @@
-    ( AnyCardanoEra (..), CardanoEra (..), FromCBOR )
+    ( FromCBOR, ToCBOR )
 import Cardano.Binary
     ( Annotator (runAnnotator)
     , FromCBOR (fromCBOR)
     , FullByteString (Full)
     , toCBOR
     )
+import Cardano.Wallet.Read.Eras
+    ( (:.:) (..)
+    , EraFun (..)
+    , EraValue
+    , K (..)
+    , applyEraFun
+    , extractEraValue
+    , sequenceEraValue
+    )
+import Cardano.Wallet.Read.Tx
+    ( Tx (..), TxT )
 import Codec.CBOR.Read
     ( DeserialiseFailure, deserialiseFromBytes )
 import Codec.CBOR.Write
     ( toLazyByteString )
-import Control.DeepSeq
-    ( NFData, rnf )
+import Data.ByteArray.Encoding
+    ( Base (Base16), convertToBase )
+import Data.ByteString.Lazy
+    ( toStrict )
 import Data.Functor.Identity
     ( Identity (..) )
-import GHC.Generics
-    ( Generic )
+import Data.Text.Class
+    ( ToText )
+import Data.Text.Encoding
+    ( decodeUtf8 )
+import Fmt
+    ( Buildable (..) )
 
-import qualified Cardano.Api as Api
-
-import qualified Cardano.Wallet.Read.Tx as Read
 import qualified Data.ByteString.Lazy as BL
 
 -- | Serialized version of a transaction. Deserializing should at least expose
 -- enough information to compute the `TxId`.
-data TxCBOR =
-    TxCBOR
-    { -- | Serialized transaction as expected from mempool on submission.
-      txCBOR :: !BL.ByteString
-      -- | Era of the transaction, to identify the right codec.
-    , txEra :: !AnyCardanoEra
-    } deriving (Show, Generic, Eq)
+type TxCBOR = EraValue (K BL.ByteString)
 
-instance Ord TxCBOR where
-    compare (TxCBOR cb _) (TxCBOR cb' _) = compare cb cb'
+instance Buildable TxCBOR where
+    build
+        = build . decodeUtf8 . convertToBase Base16 . toStrict . extractEraValue
 
-instance NFData TxCBOR where
-    rnf (TxCBOR cb _) = rnf cb -- missing instance NFData AnyCardanoEra
+instance ToText TxCBOR
 
--- | Compute the CBOR representation of a transaction.
---
--- This CBOR includes the transaction body, but also witnesses.
-getTxCBOR :: Read.Tx -> TxCBOR
-getTxCBOR (Read.Tx era tx) = TxCBOR cbor (AnyCardanoEra era)
-  where
-    -- See Note [SeeminglyRedundantPatternMatches]
-    cbor = toLazyByteString $ case era of
-        ByronEra -> toCBOR tx
-        ShelleyEra -> toCBOR tx
-        MaryEra -> toCBOR tx
-        AllegraEra -> toCBOR tx
-        AlonzoEra -> toCBOR tx
-        BabbageEra -> toCBOR tx
+-- | Render a tx into its cbor, it just applies 'serializeTx'.
+renderTxToCBOR :: EraValue Tx -> EraValue (K BL.ByteString)
+renderTxToCBOR = applyEraFun serializeTx
 
-instance FromCBOR a => FromCBOR (Identity a) where
-    fromCBOR = fmap Identity fromCBOR
+-- | CBOR serialization of a tx in any era.
+serializeTx :: EraFun Tx (K BL.ByteString)
+serializeTx = EraFun
+    { byronFun = f
+    , shelleyFun = f
+    , allegraFun = f
+    , maryFun = f
+    , alonzoFun = f
+    , babbageFun = f
+    }
+    where
+        f :: ToCBOR(TxT era) => Tx era -> K BL.ByteString  era
+        f = K . toLazyByteString . toCBOR . unTx
```

Comment:
```suggestion
  where
    f :: ToCBOR (TxT era) => Tx era -> K BL.ByteString  era
    f = K . toLazyByteString . toCBOR . unTx
```
(Formatting:
`where` 2 spaces looks nicer.

A space between constraint `ToCBOR` and its argument helps readability.
)